Utilities/Mergers/src/Merger.cxx

```diff
@@ -0,0 +1,284 @@
+
+  if (shouldPublish(ctx)) {
+
+    publish(ctx.outputs());
+
+    cleanCacheAfterPublishing();
+
+    if (mConfig.timespan.value == Timespan::LastDifference) {
+      mMergedObjects.reset();
+    }
+  }
+}
+
+std::function<void()> Merger::prepareTimerCallback(InitContext& ictx) const
+{
+  return [&timesliceIndex = ictx.services().get<TimesliceIndex>(),
+          timeLast = std::make_shared<steady_clock::time_point>(steady_clock::now()),
+          periodMs = this->mConfig.publicationDecision.param * 1000,
+          timesliceID = uint64_t(0) ]() mutable
+  {
+    auto timeNow = steady_clock::now();
+
+    if (duration_cast<milliseconds>(timeNow - *timeLast).count() > periodMs) {
+
+      data_matcher::VariableContext context;
+      context.put(data_matcher::ContextUpdate{ 0, timesliceID });
+      context.commit();
+
+      timesliceIndex.replaceLRUWith(context);
+
+      timesliceID++;
+      *timeLast = timeNow;
+    }
+  };
+}
+
+void Merger::cleanCacheAfterMerging()
+{
+  mCache.setAllMerged();
+}
+
+void Merger::cleanCacheAfterPublishing()
+{
+  if (mConfig.timespan.value == Timespan::LastDifference || mConfig.ownershipMode.value == OwnershipMode::Integral) {
+    mCache.clear();
+  }
+
+  if (mConfig.ownershipMode.value == OwnershipMode::Full) {
+    mCache.setAllMerged(false);
+  }
+  mCache.setAllUpdated(false);
+}
+
+void Merger::mergeCache()
+{
+  switch (mConfig.mergingMode.value) {
+    case MergingMode::Binwise:
+    case MergingMode::UseInterfaceFunction: {
+
+      size_t i = 0;
+      if (!mMergedObjects) {
+        for (; i < mCache.size(); i++) {
+          if (!mCache[i].deque.empty()) {
+            mMergedObjects.reset(mCache[i].deque[0].obj->Clone());
+            mCache.setMerged(i, 0);
+            break;
+          }
+        }
+      }
+
+      if (!mMergedObjects) {
+        LOG(INFO) << "mergeCache(): The cache is empty, nothing to merge.";
+        return;
+      }
+
+      auto unpackedMergedObjects = unpackObjects(mMergedObjects.get());
+      std::vector<TObjArray> unpackedCollectionsOfObjects(unpackedMergedObjects.size());
+      // todo: unpack straight to TCollection?
+      for (; i < mCache.size(); i++) {
+        for (const auto& entry : mCache[i].deque) {
+          if (!entry.is_merged) {
+            auto unpackedCachedObjects = unpackObjects(entry.obj.get());
+            assert(unpackedMergedObjects.size() == unpackedCachedObjects.size());
+
+            for (int j = 0; j < unpackedCachedObjects.size(); j++) {
+              unpackedCollectionsOfObjects[j].Add(unpackedCachedObjects[j]);
+            }
+          }
+        }
+      }
+
+      for (int k = 0; k < unpackedMergedObjects.size(); k++) {
+
+        TObject* mergedObject = unpackedMergedObjects[k];
+        const char* className = mergedObject->ClassName();
+        Long64_t errorCode = 0;
+
+        //todo: investigate -NOCHECK flag for histogram merging
+
+        if (mConfig.mergingMode.value == MergingMode::UseInterfaceFunction) {
```

Comment:
Would it not be better to have the dispatch here based on the data type and not some policy? Looks to me this should be called when an object derives from MergerInterface.

Reply:
My idea was to provide a possibility to partially implement MergeInterface, so one could f.e.  implement merge(), but leave unpack() alone. Then, usage of custom functions had to be activated in another way. 
Another option that I can think of would be to create MergeInterface, UnpackInterface and other separate interfaces. Or just force the user to implement all of the interface functions.

Reply:
i'd go for single (less complexity) and full abstract (clear responsibility). Hunting bugs at runtime due to not implemented default methods is a waste of everybody's time :)

Reply:
This is definitely a good point. On the other, I would lose a little bit of the clarity of the configuration interface. Now, the user chooses one of the options from `enum class MergingMode { Binwise, Timewise, Concatenate, UseInterfaceFunction };`. If inheriting MergeInterface quietly overwrote this choice, it would cause some confusion.
Though, I do not strongly insist on keeping it as it is.

Reply:
just drop the UseInterfaceFunction config option - it should be data driven.